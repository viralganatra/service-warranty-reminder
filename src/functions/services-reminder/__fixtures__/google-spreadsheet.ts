export const rawFixture = [
  {
    Property: 'property 1',
    ['Certificate Type']: 'cert 1',
    ['Expiry Date']: '16-May-2022',
    Link: 'https://www.google.com/1',
    _rawData: [0, 1, 2, 3],
  },
  {
    Property: null,
    ['Certificate Type']: null,
    ['Expiry Date']: null,
    Link: null,
    _rawData: [],
  },
  {
    Property: 'property 2',
    ['Certificate Type']: 'cert 2',
    ['Expiry Date']: '21-Dec-2024',
    Link: 'https://www.google.com/2',
    _rawData: [0, 1, 2, 3],
  },
  {
    Property: null,
    ['Certificate Type']: null,
    ['Expiry Date']: null,
    Link: null,
    _rawData: [],
  },
];

export const filteredFixture = [
  {
    property: 'property 1',
    certificateType: 'cert 1',
    expiryDate: '30-Apr-2022',
    linkToCertificate: 'https://www.google.com/1',
  },
  {
    property: 'property 2',
    certificateType: 'cert 2',
    expiryDate: '21-Dec-2024',
    linkToCertificate: 'https://www.google.com/2',
  },
  {
    property: 'property 3',
    certificateType: 'cert 3',
    expiryDate: '03-May-2022',
    linkToCertificate: 'https://www.google.com/3',
  },
  {
    property: 'property 4',
    certificateType: 'cert 4',
    expiryDate: '08-May-2022',
    linkToCertificate: 'https://www.google.com/4',
  },
  {
    property: 'property 5',
    certificateType: 'cert 5',
    expiryDate: '15-May-2022',
    linkToCertificate: 'https://www.google.com/5',
  },
];
