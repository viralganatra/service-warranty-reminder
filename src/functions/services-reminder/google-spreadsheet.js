const { GoogleSpreadsheet } = require('google-spreadsheet');
const getSSMParams = require('../../utils/ssm');

async function getSecrets() {
  return getSSMParams([
    'GOOGLE_SPREADSHEET_ID',
    'GOOGLE_SERVICE_ACCOUNT_EMAIL',
    'GOOGLE_PRIVATE_KEY',
  ]);
}

function spreadsheetAuth(document, secrets) {
  return document.useServiceAccountAuth({
    client_email: secrets.GOOGLE_SERVICE_ACCOUNT_EMAIL.Value,
    private_key: secrets.GOOGLE_PRIVATE_KEY.Value.replace(/\\n/gm, '\n'),
  });
}

async function getSpreadsheetData() {
  const secrets = await getSecrets();

  const doc = new GoogleSpreadsheet(secrets.GOOGLE_SPREADSHEET_ID.Value);

  await spreadsheetAuth(doc, secrets);
  await doc.loadInfo();

  const [sheet] = doc.sheetsByIndex;

  const rows = await sheet.getRows();

  return rows.map((row) => ({
    property: row.Property,
    certificateType: row['Certificate Type'],
    expiryDate: row['Expiry Date'],
    linkToCertificate: row.Link,
  }));
}

exports.getData = getSpreadsheetData;
