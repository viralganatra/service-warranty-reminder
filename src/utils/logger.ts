import { Logger } from '@aws-lambda-powertools/logger';

export const logger = new Logger({ serviceName: process.env.SERVICE_NAME });
