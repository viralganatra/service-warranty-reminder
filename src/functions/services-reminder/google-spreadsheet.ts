import { GoogleSpreadsheet, GoogleSpreadsheetRow } from 'google-spreadsheet';
import getSSMParams from '../../utils/ssm';
import { ParameterWithRequired } from '../../types/secrets';

const SSM_PARAMS = ['googleSpreadsheetId', 'googleServiceAccountEmail', 'googlePrivateKey'];

type SsmParams = 'googleSpreadsheetId' | 'googleServiceAccountEmail' | 'googlePrivateKey';

interface DataRow extends GoogleSpreadsheetRow {
  Property: string;
  'Certificate Type': string;
  'Expiry Date': string;
  Link: string;
}

function getSecrets() {
  return getSSMParams(SSM_PARAMS);
}

function spreadsheetAuth(
  document: GoogleSpreadsheet,
  secrets: Record<SsmParams, ParameterWithRequired>,
) {
  return document.useServiceAccountAuth({
    client_email: secrets.googleServiceAccountEmail?.Value,
    private_key: secrets.googlePrivateKey?.Value.replace(/\\n/gm, '\n'),
  });
}

export default async function getSpreadsheetData() {
  try {
    const secrets = await getSecrets();

    const doc = new GoogleSpreadsheet(secrets.googleSpreadsheetId?.Value);

    await spreadsheetAuth(doc, secrets);
    await doc.loadInfo();

    const [sheet] = doc.sheetsByIndex;

    const rows = (await sheet.getRows()) as DataRow[];

    return rows.map((row) => ({
      property: row.Property,
      certificateType: row['Certificate Type'],
      expiryDate: row['Expiry Date'],
      linkToCertificate: row.Link,
    }));
  } catch (err) {
    console.log(err);

    throw err;
  }
}
