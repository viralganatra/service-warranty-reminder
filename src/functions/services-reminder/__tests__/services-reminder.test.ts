import { Logger } from '@aws-lambda-powertools/logger';
import { APIGatewayProxyEvent, Context } from 'aws-lambda';
import servicesReminder from '../services-reminder';
import { filteredFixture } from '../__fixtures__/google-spreadsheet';
import sendEmail from '../../../lib/email';

jest.mock('../google-spreadsheet', () => () => filteredFixture);
jest.mock('../../../lib/email', () => jest.fn());

describe('Service reminder', () => {
  let logger: Logger;
  let proxy: APIGatewayProxyEvent;
  let context: Context;

  beforeEach(() => {
    jest.useFakeTimers();

    logger = new (jest.fn())();
    proxy = new (jest.fn())();
    context = new (jest.fn())();

    logger.addContext = jest.fn();
    logger.debug = jest.fn();
    logger.info = jest.fn();
    logger.error = jest.fn();
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
    jest.useRealTimers();
  });

  it('should add the context to the logger and log default messages', async () => {
    await servicesReminder({ logger })(proxy, context);

    expect((logger.addContext as jest.Mock).mock.calls[0][0]).toBe(context);
    expect(logger.debug).nthCalledWith(1, 'Starting check');
    expect(logger.debug).lastCalledWith('Finished check');
  });

  it('should send an email for service to notify', async () => {
    jest.setSystemTime(new Date('2022-05-01'));

    const mockSendEmail = sendEmail as jest.Mock;
    const result = await servicesReminder({ logger })(proxy, context);

    expect(logger.info).lastCalledWith('Sending email for Property: property 5 cert 5 15-May-2022');
    expect(logger.info).toBeCalledTimes(4);
    expect(result).toMatchInlineSnapshot(`
      Object {
        "body": "{
        \\"message\\": \\"Ok\\"
      }",
        "statusCode": 200,
      }
    `);

    expect(mockSendEmail).toBeCalledTimes(4);
    expect(mockSendEmail.mock.calls[0][0]).toMatchSnapshot();
    expect(mockSendEmail.mock.calls[1][0]).toMatchSnapshot();
  });

  it('should not send an email if there are no services to notify', async () => {
    jest.setSystemTime(new Date('2023-01-01'));

    const result = await servicesReminder({ logger })(proxy, context);

    expect(logger.info).lastCalledWith('No emails to send');
    expect(result).toMatchInlineSnapshot(`
      Object {
        "body": "{
        \\"message\\": \\"No emails to send\\"
      }",
        "statusCode": 200,
      }
    `);
  });

  it('should throw an error if the data schema is invalid', async () => {
    jest.doMock('../google-spreadsheet', () => ({
      __esModule: true,
      default: () => [{ property: 'prop 1' }],
    }));

    const mockSendEmail = jest.fn();

    jest.doMock('../../../lib/email', () => ({
      __esModule: true,
      default: mockSendEmail,
    }));

    const localServicesReminder = await import('../services-reminder');
    const result = await localServicesReminder.default({ logger })(proxy, context);

    expect(result).toEqual({
      statusCode: 500,
      body: '"InvalidSchemaError: ValidationError: [0].linkToCertificate is a required field"',
    });

    expect(logger.error).toBeCalled();
    expect(mockSendEmail).toBeCalledWith(
      expect.objectContaining({
        subject: 'Error with property service warranty',
        html: expect.any(String),
      }),
    );
  });
});
