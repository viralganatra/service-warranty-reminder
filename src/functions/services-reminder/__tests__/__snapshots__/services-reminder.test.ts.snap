// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service reminder should send an email for service to notify 2`] = `
Object {
  "html": "
  <!DOCTYPE html>
  <html lang=\\"en\\" xmlns:v=\\"urn:schemas-microsoft-com:vml\\" xmlns:o=\\"urn:schemas-microsoft-com:office:office\\">
    <head>
      <meta charset=\\"utf-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\">
      <meta name=\\"format-detection\\" content=\\"telephone=no, date=no, address=no, email=no, url=no\\">
      <meta name=\\"x-apple-disable-message-reformatting\\">
      <meta name=\\"color-scheme\\" content=\\"light dark\\">
      <meta name=\\"supported-color-schemes\\" content=\\"light dark\\">
      <!--[if mso]>
      <noscript>
        <xml>
          <o:OfficeDocumentSettings>
            <o:PixelsPerInch>96</o:PixelsPerInch>
          </o:OfficeDocumentSettings>
        </xml>
      </noscript>
      <![endif]-->
      <style>
        :root {
          color-scheme: light dark;
          supported-color-schemes: light dark;
        }

        .body,
        .container {
          font-family: sans-serif;
          color: #39444d;
        }

        .container {
          padding: 30px;
        }

        h1 {
          font-size: 32px;
          font-weight: 600;
          margin: 0 0 20px 0;
        }

        h2 {
          font-size: 24px;
          font-weight: 400;
          margin: 24px 0 8px 0;
        }

        h3 {
          font-size: 18px;
          font-weight: 600;
        }

        p {
          font-size: 16px;
          line-height: 24px;
          margin: 0 0 24px 0;
        }

        hr {
          border: 1px solid #c5ced6;
          margin: 48px 0 48px 0;
        }

        a,
        a:visited {
          color: #1273e6;
        }

        .logo {
          margin: 0 0 58px 0;
        }

        a.social,
        a.social:visited {
          font-size: 14px;
          color: #626f7a;
        }
      </style>
    </head>
    <body class=\\"body\\">
      <div style=\\"font-size:16px; font-size:1rem;\\">
        <div class=\\"container\\">
          <!--[if true]>
          <table role=\\"presentation\\" width=\\"100%\\" align=\\"center\\" style=\\"background:#ffffff\\">
            <tr>
              <td></td>
              <td style=\\"width:600px;background:#ffffff;padding:24px;\\">
          <![endif]-->
                <div style=\\"max-width:600px; margin:0 auto;background:#ffffff;padding:0 24px;mso-padding-alt:0;\\">
                
    
      <h1>Warranty Expired!</h1>
      <p>The cert 1 warranty for property 1 expired on 30-Apr-2022.</p>
    
    <p>Please take action.</p>
    <hr />
    <h2>What to do</h2>
    <p>You can <a href=\\"https://www.google.com/1\\">view the current certificate</a></p>

    <p>
      Please contact the relevant party to renew the warranty or ignore this message if you do not wish to take action.
    </p>
  
                </div>
              <!--[if true]>
              </td>
            </tr>
          </table>
          <![endif]-->
          </div>
        </div>
      </body>
    </html>
  ",
  "subject": "cert 1 warranty expired for property 1 - please take action",
}
`;

exports[`Service reminder should send an email for service to notify 3`] = `
Object {
  "html": "
  <!DOCTYPE html>
  <html lang=\\"en\\" xmlns:v=\\"urn:schemas-microsoft-com:vml\\" xmlns:o=\\"urn:schemas-microsoft-com:office:office\\">
    <head>
      <meta charset=\\"utf-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\">
      <meta name=\\"format-detection\\" content=\\"telephone=no, date=no, address=no, email=no, url=no\\">
      <meta name=\\"x-apple-disable-message-reformatting\\">
      <meta name=\\"color-scheme\\" content=\\"light dark\\">
      <meta name=\\"supported-color-schemes\\" content=\\"light dark\\">
      <!--[if mso]>
      <noscript>
        <xml>
          <o:OfficeDocumentSettings>
            <o:PixelsPerInch>96</o:PixelsPerInch>
          </o:OfficeDocumentSettings>
        </xml>
      </noscript>
      <![endif]-->
      <style>
        :root {
          color-scheme: light dark;
          supported-color-schemes: light dark;
        }

        .body,
        .container {
          font-family: sans-serif;
          color: #39444d;
        }

        .container {
          padding: 30px;
        }

        h1 {
          font-size: 32px;
          font-weight: 600;
          margin: 0 0 20px 0;
        }

        h2 {
          font-size: 24px;
          font-weight: 400;
          margin: 24px 0 8px 0;
        }

        h3 {
          font-size: 18px;
          font-weight: 600;
        }

        p {
          font-size: 16px;
          line-height: 24px;
          margin: 0 0 24px 0;
        }

        hr {
          border: 1px solid #c5ced6;
          margin: 48px 0 48px 0;
        }

        a,
        a:visited {
          color: #1273e6;
        }

        .logo {
          margin: 0 0 58px 0;
        }

        a.social,
        a.social:visited {
          font-size: 14px;
          color: #626f7a;
        }
      </style>
    </head>
    <body class=\\"body\\">
      <div style=\\"font-size:16px; font-size:1rem;\\">
        <div class=\\"container\\">
          <!--[if true]>
          <table role=\\"presentation\\" width=\\"100%\\" align=\\"center\\" style=\\"background:#ffffff\\">
            <tr>
              <td></td>
              <td style=\\"width:600px;background:#ffffff;padding:24px;\\">
          <![endif]-->
                <div style=\\"max-width:600px; margin:0 auto;background:#ffffff;padding:0 24px;mso-padding-alt:0;\\">
                
    
      <h1>Warranty Expiring!</h1>
      <p>The cert 3 warranty for property 3 is expiring on 03-May-2022 (in 2 days time).</p>
    
    <p>Please take action.</p>
    <hr />
    <h2>What to do</h2>
    <p>You can <a href=\\"https://www.google.com/3\\">view the current certificate</a></p>

    <p>
      Please contact the relevant party to renew the warranty or ignore this message if you do not wish to take action.
    </p>
  
                </div>
              <!--[if true]>
              </td>
            </tr>
          </table>
          <![endif]-->
          </div>
        </div>
      </body>
    </html>
  ",
  "subject": "cert 3 warranty expiring for property 3 - please take action",
}
`;
